### PR TITLE
In-memory engine: gc the range immediately after load

### LINKED ISSUE #17299: In-memory engine: gc the range before completing load
## Development Task
Currently, range load all keys in rocksdb for the range, and we may generate a safe point and only load the keys that can be read.

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #17299

What's Changed:

```commit-message
gc the range immediately after load 
```
This PR optimizes the load by gc the range immediately after load. We can filter the key during the load rather than load then gc, but it can bring more complexities. I leave it as a TODO.

The effect of this PR:
image

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/range_cache_memory_engine/src/background.rs [hunk 1]
```diff
@@ -956,34 +970,90 @@ impl Runnable for BackgroundRunner {
...
                                         error!("creating rocksdb iterator failed"; "cf" => cf, "err" => %e);
-                                        return false;
+                                        return None;
                                     }
                                 }
                             }
-                            true
+
+                            // gc the range
+                            let tso_timeout = std::cmp::min(gc_interval, Duration::from_secs(5));
```
Comment (LykxSassinator) [inline components/range_cache_memory_engine/src/background.rs:979]: Please add more comments / annotations on the default value `5 seconds` and make it as a constant value.

File: components/range_cache_memory_engine/src/background.rs [hunk 2]
```diff
@@ -956,34 +970,90 @@ impl Runnable for BackgroundRunner {
...
+                            let now = match block_on_timeout(pd_client.get_tso(), tso_timeout) {
+                                Ok(Ok(ts)) => ts,
+                                err => {
+                                    error!(
+                                        "get timestamp failed, skip gc loaded range";
+                                        "timeout_duration" => ?tso_timeout,
+                                        "error" => ?err,
+                                    );
+                                    // Get timestamp fail so don't do gc.
+                                    return Some(0);
```
Comment (LykxSassinator) [inline components/range_cache_memory_engine/src/background.rs:989]: If so => `now == Some(0)`, maybe all following steps can be skipped with directly returning `None` ?

File: components/range_cache_memory_engine/src/background.rs [hunk 3]
```diff
@@ -956,34 +970,90 @@ impl Runnable for BackgroundRunner {
...
+                                iter.next(guard);
+                            }
+
+                            Some(safe_point)
                         };
 
                         let start = Instant::now();
-                        if !snapshot_load() {
+                        if let Some(safe_point) = snapshot_load() {
+                            if core.on_snapshot_load_finished(
```
Comment (LykxSassinator) [inline components/range_cache_memory_engine/src/background.rs:1034]:
```suggestion
                        if let Some(safe_point) = snapshot_load()
                            && core.on_snapshot_load_finished(
```

File: components/range_cache_memory_engine/src/background.rs [hunk 4]
```diff
@@ -956,34 +970,90 @@ impl Runnable for BackgroundRunner {
...
+                            let mut iter = write_cf_handle.iterator();
+                            let guard = &epoch::pin();
+                            let (start_key, end_key) = encode_key_for_boundary_with_mvcc(&range);
+                            iter.seek(&start_key, guard);
+                            while iter.valid() && iter.key() < &end_key {
+                                let k = iter.key();
+                                let v = iter.value();
+                                if let Err(e) = filter.filter(k.as_bytes(), v.as_bytes()) {
+                                    warn!(
+                                        "Something Wrong in memory engine GC";
```
Comment (overvenus) [inline components/range_cache_memory_engine/src/background.rs:1022]:
```suggestion
                                        "something wrong in memory engine GC";
```

Comment (overvenus) [reply components/range_cache_memory_engine/src/background.rs:1022]: How about evicting the range if it occurs error?

File: components/range_cache_memory_engine/src/background.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/range_cache_memory_engine/src/background.rs:989]: None means load is failed. Some(0) means the load success but does not get the tso for gc.

File: components/range_cache_memory_engine/src/background.rs (hunk 4 above)
Comment (SpadeA-Tang) [reply components/range_cache_memory_engine/src/background.rs:1022]: It follows what what compaction filter does.

File: components/range_cache_memory_engine/src/background.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/range_cache_memory_engine/src/background.rs]: Done

File: components/range_cache_memory_engine/src/background.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/range_cache_memory_engine/src/background.rs:1036]: why this?

Comment (LykxSassinator) [reply components/range_cache_memory_engine/src/background.rs:1036]: Why not directly use the extracted `safe_point` from the `Some(safe_point)` to do the later judgement ?

File: components/range_cache_memory_engine/src/background.rs (hunk 2 above)
Comment (LykxSassinator) [reply components/range_cache_memory_engine/src/background.rs:991]: got

File: components/range_cache_memory_engine/src/background.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/range_cache_memory_engine/src/background.rs:1036]: I still do not understand what you mean, You seem to merge two if but there's a else for the second if. By the way, in on_snapshot_load_finished, it still can fail.

**Review by overvenus — COMMENTED** [review]
Rest LGTM

**Review by overvenus — APPROVED** [review]

**Review by LykxSassinator — APPROVED** [review]

**Pull request merged**